Complete console scripts installed from wheels

evanunderscore opened the pull request:
Fixes #225 

Probably needs some tests, but I'm putting this up for discussion and so I don't forget about it.

mbdevpl commented:
@evanunderscore With latest release 1.10.0 the global completion for console scripts is still not working because of missing `# EASY-INSTALL-...` comment, as described in #225. And pip maintainers say in https://github.com/pypa/pip/issues/4702 that they will not do anything about it. So, I rebased this PR onto latest master (1.10.0), and now it's working! Thanks!

In my case I'm using pyenv.

I'm not sure why CI tests are failing, but in practice this PR definitely works, so it must be some corner case which doesn't occur in my environment. Is there any chance for the tests to be fixed/added and for merging this into master at some point?

evanunderscore commented:
The CI failure seems to be something unrelated on OSX, and codecov is complaining that I haven't tested enough. I'm short on free time, but I can try to get this up to scratch and merged soon.

evanunderscore commented:
@mbdevpl Can you verify this still works for you? I've made some very minor changes in the process of bringing this up to speed with master.

@kislyuk Would you like to review this? I don't believe we have any tests specifically targeting `python-argcomplete-check-easy-install-script` so I'd appreciate if you could sanity check that I haven't obviously broken that.

evanunderscore commented:
Unfortunately it looks like this OSX failure *is* related to the tests I've written, so I'll have to dig into that and see what the problem is.

evanunderscore commented:
@kislyuk Thanks for reviewing. I haven't managed to look at the failure yet but I'll report back as soon as I do.

evanunderscore commented:
@kislyuk I've managed to fix the issue, but it requires some fairly specific workarounds. I don't fully understand the problem but tried to link the related issues where relevant. If you're happy with this, it's good to go.

kislyuk commented on argcomplete/_check_console_script.py:
```diff
@@ -0,0 +1,42 @@
+    script_path = sys.argv[1]
+
+    # Find the module and function names that correspond to this
+    # assuming it as actually a console script.
```
"assuming it is"

evanunderscore replied:
Addressed in #263

kislyuk commented on argcomplete/_check_console_script.py:
```diff
@@ -0,0 +1,42 @@
+import os
```
It would be nice to add a docstring here describing what this file does and linking to the definition and docs for console scripts.

evanunderscore replied:
Addressed in #263

evanunderscore commented:
Thanks for looking into this @kislyuk. I'll let you know when I've fixed up the review comments and pushed them to master.

> What happens if the user is running a Python that does not have pkg_resources available

Whoops, I'd assumed `pkg_resources` was built-in, but it looks like it's included with `setuptools`. If it's missing, the script will fail with an `ImportError` that will likely trigger automated error reporting we addressed in #261. However if we think this will only happen if the package is faulty, then that's probably for the best. Thoughts?

> What kind of worst-case real-life performance impact we can expect to see here, given that this script is going into the default code path for global completion

There are some minor performance implications related to the reordering of file accesses, but the important one is running the script. If tab completion is attempted on a Python script that is *not* meant to be tab completed via `argcomplete` one way or another, you'll end up down this branch, and you'll pay the performance cost of running the interpreter (and new script). However as touched on [here](https://github.com/kislyuk/argcomplete#common-problems), if you fail to complete any name, a default completer will get registered, so any subsequent attempts to tab complete that same name will have no penalty.
